### PR TITLE
Check for empty BAM

### LINKED ISSUE #47: Empty BAM check
With the current `samtools quickcheck`, properly formatted empty BAM files pass the validation. We want to add a check to catch these empty BAMs and an option/flag for the tool to enable/disable this check

### PR DESCRIPTION
**Pull request by aybeshlikyan**
### Formatting

- [X] I have read the [code review guidelines](https://confluence.mednet.ucla.edu/display/BOUTROSLAB/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://confluence.mednet.ucla.edu/display/BOUTROSLAB/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [X] The name of the branch is meaningful and well formatted following the [standards](https://confluence.mednet.ucla.edu/display/BOUTROSLAB/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].

- [X] I have set up or verified the branch protection rule following the [github standards](https://confluence.mednet.ucla.edu/pages/viewpage.action?spaceKey=BOUTROSLAB&title=GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

### File Updates

- [X] I have ensured that the version number update follows the [versioning standards](https://confluence.mednet.ucla.edu/display/BOUTROSLAB/Docker+image+versioning+standardization).

- [X] I have updated the version number/dependencies and added my name to the maintainer list in the `Dockerfile`.

- [X] I have updated the version number/feature changes in the `README.md`.

- [X] I have updated the version number and added my name to the contributors list in the `metadata.yaml`.

- [X] I have added the changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

- [X] I have drafted the new version release with any additions/changes and have linked the `CHANGELOG.md` in the release. 

### Docker Hub Auto Build Rules

- [X] I have created automated build rules following [this page](https://confluence.mednet.ucla.edu/display/BOUTROSLAB/How+to+set+up+automated+builds+for+Docker+Hub) and I have not manually pushed this Docker image to the `blcdsdockerregistry` on [Docker Hub](https://hub.docker.com).

### Docker Image Testing

- [X] I have tested the Docker image with the `docker run` command as described below.

#### Test the Docker image with at least one sample. Verify the new Docker image works using:

```docker run -u $(id -u):$(id -g) –w <working-directory> -v <directory-you-want-to-mount>:<how-you-want-to-mount-it-within-the-docker> --rm <docker-image-name> <command-to-the-docker-with-all-parameters>```

#### My command: 

```docker build -t test-pipeval . -f docker/Dockerfile```
```docker run -it --rm -v /Users/abeshlikyan/Infrastructure/public-tool-PipeVal/tests:/tool/tests test-pipeval pytest```
```docker run -it --rm -v /Users/abeshlikyan/Infrastructure/public-tool-PipeVal/tests:/tool/tests test-pipeval pytest tests/test_main.py::test_file_bam_empty```

*Also tested manually on* `/hot/resource/SMC-HET/normal/bams/A-null/HG002_N_A-null.bam`

## Description

Closes #47. Adds check for BAM file with valid header but no reads.

### CODE REVIEW

**Comment by tyamaguchi-ucla** [comment]
> _Also tested manually on_ `/hot/resource/SMC-HET/normal/bams/A-null/HG002_N_A-null.bam`

Great! Can you show the output of this test and A-mini n2?

File: validate/validators/bam.py [hunk 1]
```diff
@@ -9,4 +9,8 @@ def validate_bam_file(path):
     except pysam.SamtoolsError as err:
         raise ValueError("samtools bam check failed. " + str(err)) from err
 
+    num_lines = int(pysam.view("-c", str(path)).strip())
+    if num_lines <= 0:
+        raise ValueError("pysam bam check failed. No reads in " + str(path))
```
Comment (yashpatel6) [inline validate/validators/bam.py:14]: For large BAMs, running the full `samtools view -c` command might take significant time, it would be worth looking into modifying the check to something like `samtools view --no-header <bam> | head -n 1` and checking if what's returned isn't an empty string. Pysam has the [AlignmentFile](https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignmentFile) class, which has a `head` method which might prove useful for doing this.

Comment (tyamaguchi-ucla) [reply validate/validators/bam.py:14]: >Pysam has the [AlignmentFile](https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignmentFile) class, which has a head method which might prove useful for doing this.

@aybeshlikyan did you get a chance to try the method? https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignmentFile.head

File: validate/validators/bam.py [hunk 2]
```diff
@@ -9,4 +9,8 @@ def validate_bam_file(path):
     except pysam.SamtoolsError as err:
         raise ValueError("samtools bam check failed. " + str(err)) from err
 
+    bam_it = pysam.AlignmentFile(str(path)).head(1)
```
Comment (yashpatel6) [inline validate/validators/bam.py:12]: In think generally, we want to avoid having the type of a variable in the variable name, right @tyamaguchi-ucla @aholmes?

Comment (aholmes) [reply validate/validators/bam.py:12]: It depends, however I do personally try to avoid it. Here, I would use a type hint. The pysam docs say it is a `IteratorRowHead`, so something like this, perhaps.

```python
bam_head: IteratorRowHead = pysam.AlignmentFile(str(path)).head(1)
```

File: tests/test_main.py [hunk 3]
```diff
@@ -81,6 +81,17 @@ def test_path_writable_io_error(mock_os_access):
     with pytest.raises(IOError):
         validate.__main__.path_writable(test_path)
 
+@mock.patch('validate.validators.bam.pysam.quickcheck')
+@mock.patch('validate.validators.bam.pysam.view')
+def test_file_bam_empty(mock_pysam_quickcheck, mock_pysam_view):
+    '''Tests Value error for valid BAM header with no reads'''
+    mock_pysam_quickcheck.return_value = ''
+    mock_pysam_view.return_value = '0\n'
+    test_path = Path('empty/valid/bam')
```
Comment (aholmes) [inline tests/test_main.py:90]: You might consider mocking `Path` as well, so any operations performed on it do not actually touch a filesystem.

Comment (aybeshlikyan) [reply tests/test_main.py:91]: That makes sense! It's used in older unit tests as well so I can clean up the whole file in a separate dedicated PR.

**Review by aholmes — APPROVED** [review]

**Review by tyamaguchi-ucla — APPROVED** [review]
Looks good to me. Anything else to add @yashpatel6 ?

**Review by yashpatel6 — APPROVED** [review]
Looks good!
